File: projects/java/cdm/src/mil/tatrc/physiology/datamodel/scenario/datarequests/SEDataRequestManager.java
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/
package mil.tatrc.physiology.datamodel.scenario.datarequests;

import java.lang.reflect.Constructor;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;

import mil.tatrc.physiology.datamodel.bind.DataRequestData;
import mil.tatrc.physiology.datamodel.CDMSerializer;
import mil.tatrc.physiology.datamodel.bind.DataRequestsData;
import mil.tatrc.physiology.datamodel.substance.SESubstance;
import mil.tatrc.physiology.datamodel.substance.SESubstanceManager;
import mil.tatrc.physiology.utilities.Log;

public class SEDataRequestManager
{
	protected String                        filename;
	// TODO Decimal Formatting Data
	protected List<SEDataRequest>           requestedData;	
	protected double                        samplesPerSecond;
	
	public SEDataRequestManager()
	{
		this.filename = "";
		this.samplesPerSecond = 0;
		this.requestedData = new ArrayList<SEDataRequest>();
	}
	
	public void reset() 
	{
		this.filename = "";
	  this.samplesPerSecond = 0;
		requestedData.clear();
	}
	
	public boolean load(DataRequestsData in, SESubstanceManager subMgr)
	{
		reset();
		
		if(in.getFilename()!=null)
			this.filename = in.getFilename();
		if(in.getSamplesPerSecond()!=null)
			this.samplesPerSecond = in.getSamplesPerSecond();
		
		try
		{
			for (DataRequestData dataRequestData : in.getDataRequest())
			{
				String className = dataRequestData.getClass().getName().replaceAll("bind.", "scenario.datarequests.SE");
				className = className.substring(0, className.lastIndexOf("Data"));
				Class<?> klass = Class.forName(className);
				
				Object instance = null;
                for(Constructor<?> ctor : klass.getConstructors())
                {
                    Class<?>[] params = ctor.getParameterTypes();
                    if(params.length>0)
                    {
                        if(params[0].isAssignableFrom(SESubstance.class))
                        {
                            Method m=dataRequestData.getClass().getMethod("getSubstance");
                            String subName = (String)m.invoke(dataRequestData);
                            instance=ctor.newInstance(subMgr.getSubstance(subName));
                        }
                    }
                }
                if(instance==null)
                    instance = klass.newInstance();
				
				// Find the right "load" method
				Method loadMethod = null;
				for (Method m : klass.getMethods())
				{
					if  (m.getName().equals("load"))
					{
						loadMethod = m;
						break;
					}
				}
				if (loadMethod.getParameterTypes().length == 2)
					loadMethod.invoke(instance, dataRequestData, subMgr);
				else
					loadMethod.invoke(instance, dataRequestData);
				
				this.requestedData.add((SEDataRequest) instance);
			}
		}
		catch (Exception e)
		{
			Log.error("Unsupported Data Request received in scenario file", e);
			return false;
		}
		return true;
	}
	
	public DataRequestsData unload()
	{
		DataRequestsData data = CDMSerializer.objFactory.createDataRequestsData();
		unload(data);
		return data;
	}
	protected void unload(DataRequestsData data)
	{
		if(this.hasFilename())
			data.setFilename(this.filename);
		if(this.samplesPerSecond>0)
			data.setSamplesPerSecond(this.samplesPerSecond);
		for(SEDataRequest dr : this.requestedData)
			data.getDataRequest().add(dr.unload());
	}
	
	public boolean hasFilename()
	{
		return this.filename!=null&&!this.filename.isEmpty();
	}
	public void setFilename(String f)
	{
		this.filename = f;
	}
	public String getFilename()
	{
		return this.filename;
	}
	
	public void setSamplesPerSecond(double d)
	{
		this.samplesPerSecond=d;
	}
	public double getSamplesPerSecond()
	{
		return this.samplesPerSecond;
	}
	
	public List<SEDataRequest> getRequestedData() 
	{
		return requestedData;
	}
	
}

File: projects/biogears/libBiogears/src/cdm/scenario/requests/SEDataRequestManager.cpp
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/
#include <biogears/cdm/scenario/requests/SEDataRequestManager.h>

#include "io/cdm/DataRequests.h"

#include <biogears/cdm/Serializer.h>
#include <biogears/cdm/substance/SESubstanceManager.h>
#include <biogears/cdm/utils/unitconversion/CompoundUnit.h>
#include <biogears/engine/BioGearsPhysiologyEngine.h>
#include <biogears/filesystem/path.h>

namespace biogears {
SEDataRequestManager::SEDataRequestManager(Logger* logger)
  : Loggable(logger)
  , m_SamplesPerSecond(1.0)
  , m_ResultsFile("")

{
  m_DefaultDecimalFormatting = nullptr;
  m_OverrideDecimalFormatting = nullptr;
}
//-----------------------------------------------------------------------------
SEDataRequestManager::~SEDataRequestManager()
{
  Clear();
}
//-----------------------------------------------------------------------------
const char* SEDataRequestManager::GetResultsFilename_cStr() const
{
  return m_ResultsFile.c_str();
}
//-----------------------------------------------------------------------------
std::string SEDataRequestManager::GetResultsFilename() const
{
  return m_ResultsFile;
}
//-----------------------------------------------------------------------------
void SEDataRequestManager::SetResultsFilename(const char* name)
{
  m_ResultsFile = name;
}
//-----------------------------------------------------------------------------
void SEDataRequestManager::SetResultsFilename(const std::string& name)
{
  m_ResultsFile = name;
}
//-----------------------------------------------------------------------------
std::string SEDataRequestManager::GetWorkingDir() const
{
  return m_WorkingDir;
}
//-----------------------------------------------------------------------------
const char* SEDataRequestManager::GetWorkingDir_cStr() const
{
  return m_WorkingDir.c_str();
}
//-----------------------------------------------------------------------------
void SEDataRequestManager::SetWorkingDir(const char* name)
{
  m_WorkingDir = name;
}
//-----------------------------------------------------------------------------
void SEDataRequestManager::SetWorkingDir(const std::string& name)
{
  m_WorkingDir = name;
}
//-----------------------------------------------------------------------------
std::string SEDataRequestManager::GetResovedFilePath() const { return m_WorkingDir + m_ResultsFile; }
//-----------------------------------------------------------------------------
void SEDataRequestManager::Clear()
{
  m_SamplesPerSecond = 1.0; // Sample every time step
  DELETE_VECTOR(m_Requests);
  SAFE_DELETE(m_DefaultDecimalFormatting);
  SAFE_DELETE(m_OverrideDecimalFormatting);
}
//-----------------------------------------------------------------------------
bool SEDataRequestManager::HasDefaultDecimalFormatting() const
{
  return m_DefaultDecimalFormatting != nullptr;
}
//-----------------------------------------------------------------------------
SEDecimalFormat& SEDataRequestManager::GetDefaultDecimalFormatting()
{
  if (m_DefaultDecimalFormatting == nullptr)
    m_DefaultDecimalFormatting = new SEDecimalFormat();
  return *m_DefaultDecimalFormatting;
}
//-----------------------------------------------------------------------------
void SEDataRequestManager::RemoveDefaultDecimalFormatting()
{
  SAFE_DELETE(m_DefaultDecimalFormatting);
}
//-----------------------------------------------------------------------------
bool SEDataRequestManager::HasOverrideDecimalFormatting() const
{
  return m_OverrideDecimalFormatting != nullptr;
}
//-----------------------------------------------------------------------------
SEDecimalFormat& SEDataRequestManager::GetOverrideDecimalFormatting()
{
  if (m_OverrideDecimalFormatting == nullptr)
    m_OverrideDecimalFormatting = new SEDecimalFormat();
  return *m_OverrideDecimalFormatting;
}
//-----------------------------------------------------------------------------
void SEDataRequestManager::RemoveOverrideDecimalFormatting()
{
  SAFE_DELETE(m_OverrideDecimalFormatting);
}
//-----------------------------------------------------------------------------
void SEDataRequestManager::CreateDataRequest(std::unique_ptr<SEDataRequest> request)
{
  m_Requests.push_back(request.release());
}
//-----------------------------------------------------------------------------
SEEnvironmentDataRequest& SEDataRequestManager::CreateEnvironmentDataRequest(const SEDecimalFormat* dfault)
{
  SEEnvironmentDataRequest* dr = new SEEnvironmentDataRequest(dfault);
  m_Requests.push_back(dr);
  return *dr;
}
//-----------------------------------------------------------------------------
SEEquipmentDataRequest& SEDataRequestManager::CreateEquipmentDataRequest(const SEDecimalFormat* dfault)
{
  SEEquipmentDataRequest* dr = new SEEquipmentDataRequest(dfault);
  m_Requests.push_back(dr);
  return *dr;
}
//-----------------------------------------------------------------------------
SEPatientDataRequest& SEDataRequestManager::CreatePatientDataRequest(const SEDecimalFormat* dfault)
{
  SEPatientDataRequest* dr = new SEPatientDataRequest(dfault);
  m_Requests.push_back(dr);
  return *dr;
}
//-----------------------------------------------------------------------------
SEPhysiologyDataRequest& SEDataRequestManager::CreatePhysiologyDataRequest(const SEDecimalFormat* dfault)
{
  SEPhysiologyDataRequest* dr = new SEPhysiologyDataRequest(dfault);
  m_Requests.push_back(dr);
  return *dr;
}
//-----------------------------------------------------------------------------
SESubstanceDataRequest& SEDataRequestManager::CreateSubstanceDataRequest(const SEDecimalFormat* dfault)
{
  SESubstanceDataRequest* dr = new SESubstanceDataRequest(dfault);
  m_Requests.push_back(dr);
  return *dr;
}
//-----------------------------------------------------------------------------
SEGasCompartmentDataRequest& SEDataRequestManager::CreateGasCompartmentDataRequest(const SEDecimalFormat* dfault)
{
  SEGasCompartmentDataRequest* dr = new SEGasCompartmentDataRequest(dfault);
  m_Requests.push_back(dr);
  return *dr;
}
//-----------------------------------------------------------------------------
SELiquidCompartmentDataRequest& SEDataRequestManager::CreateLiquidCompartmentDataRequest(const SEDecimalFormat* dfault)
{
  SELiquidCompartmentDataRequest* dr = new SELiquidCompartmentDataRequest(dfault);
  m_Requests.push_back(dr);
  return *dr;
}
//-----------------------------------------------------------------------------
SEThermalCompartmentDataRequest& SEDataRequestManager::CreateThermalCompartmentDataRequest(const SEDecimalFormat* dfault)
{
  SEThermalCompartmentDataRequest* dr = new SEThermalCompartmentDataRequest(dfault);
  m_Requests.push_back(dr);
  return *dr;
}
//-----------------------------------------------------------------------------
SETissueCompartmentDataRequest& SEDataRequestManager::CreateTissueCompartmentDataRequest(const SEDecimalFormat* dfault)
{
  SETissueCompartmentDataRequest* dr = new SETissueCompartmentDataRequest(dfault);
  m_Requests.push_back(dr);
  return *dr;
}
//
bool SEDataRequestManager::DuplicateRequest(SEDataRequest* request)
{
  bool duplicate = false;
  for (auto& req : m_Requests) {
    if (request->HashCode() == req->HashCode()) {
      duplicate = true;
      break;
    }
  }
  return duplicate;
}

//-----------------------------------------------------------------------------
bool SEDataRequestManager::HasResultsFilename() const
{
  return !m_ResultsFile.empty();
}
//-----------------------------------------------------------------------------
bool SEDataRequestManager::HasWorkingDir() const
{
  return !m_ResultsFile.empty();
}
//-----------------------------------------------------------------------------
double SEDataRequestManager::GetSamplesPerSecond() const
{
  return m_SamplesPerSecond;
}
//-----------------------------------------------------------------------------
void SEDataRequestManager::SetSamplesPerSecond(double num)
{
  m_SamplesPerSecond = num;
}
//-----------------------------------------------------------------------------
bool SEDataRequestManager::HasDataRequests() const
{
  return !m_Requests.empty();
}
//-----------------------------------------------------------------------------
const std::vector<SEDataRequest*>& SEDataRequestManager::GetDataRequests()
{
  return m_Requests;
}
//-----------------------------------------------------------------------------
bool SEDataRequestManager::operator==(SEDataRequestManager const& rhs) const
{
  if (this == &rhs)
    return true;

  return m_SamplesPerSecond == rhs.m_SamplesPerSecond
    && m_Requests == rhs.m_Requests
    && m_ResultsFile == rhs.m_ResultsFile
    && m_WorkingDir == rhs.m_WorkingDir
    && ((m_DefaultDecimalFormatting && rhs.m_DefaultDecimalFormatting) ? m_DefaultDecimalFormatting->operator==(*rhs.m_DefaultDecimalFormatting) : m_DefaultDecimalFormatting == rhs.m_DefaultDecimalFormatting)
    && ((m_OverrideDecimalFormatting && rhs.m_OverrideDecimalFormatting) ? m_OverrideDecimalFormatting->operator==(*rhs.m_OverrideDecimalFormatting) : m_OverrideDecimalFormatting == rhs.m_OverrideDecimalFormatting);
}
bool SEDataRequestManager::operator!=(SEDataRequestManager const& rhs) const
{
  return !(*this == rhs);
}
//-----------------------------------------------------------------------------
}